// vue-pdf-embed/dist/index.essential.mjs
import { getDocument as X, PasswordResponses as Y, AnnotationLayer as J, TextLayer as K } from "pdfjs-dist/legacy/build/pdf.mjs";
import { GlobalWorkerOptions as ke } from "pdfjs-dist/legacy/build/pdf.mjs";
import { shallowRef as _, watchEffect as Q, toValue as I, watch as B, onBeforeUnmount as H, defineComponent as Z, ref as ee, toRef as te, computed as ae, openBlock as C, createElementBlock as A, Fragment as ne, renderList as oe, renderSlot as O, createElementVNode as z, normalizeStyle as ie, createCommentVNode as j } from "vue";
import { PDFLinkService as se } from "pdfjs-dist/legacy/web/pdf_viewer.mjs";
function re(o, c, r) {
  const t = o.contentWindow.document.createElement("style");
  t.textContent = `
    @page {
      margin: 3mm;
      size: ${c}pt ${r}pt;
    }
    body {
      margin: 0;
    }
    canvas {
      width: 100%;
      page-break-before: avoid;
      break-before: avoid;
      page-break-inside: avoid;
      break-inside: avoid;
    }
    canvas:not(:last-child) {
      page-break-after: always;
      break-after: page;
    }
  `, o.contentWindow.document.head.appendChild(t), o.contentWindow.document.body.style.width = "100%";
}
function le(o) {
  return new Promise((c) => {
    const r = document.createElement("iframe");
    r.width = "0", r.height = "0", r.style.position = "absolute", r.style.top = "0", r.style.left = "0", r.style.border = "none", r.style.overflow = "hidden", r.onload = () => c(r), o.appendChild(r);
  });
}
function de(o, c) {
  const r = URL.createObjectURL(
    new Blob([o], {
      type: "application/pdf"
    })
  ), t = document.createElement("a");
  t.href = r, t.download = c, t.style.display = "none", document.body.append(t), t.click(), setTimeout(() => {
    URL.revokeObjectURL(r), document.body.removeChild(t);
  }, 1e3);
}
function q(o) {
  for (; o != null && o.firstChild; )
    o.removeChild(o.firstChild);
}
function D(o) {
  return o ? Object.prototype.hasOwnProperty.call(o, "_pdfInfo") : !1;
}
function N(o) {
  var c;
  o.width = 1, o.height = 1, (c = o.getContext("2d")) == null || c.clearRect(0, 0, 1, 1);
}
function V(o) {
  o == null || o.querySelectorAll("canvas").forEach(N);
}
function ce({
  onError: o,
  onPasswordRequest: c,
  onProgress: r,
  source: t
}) {
  const l = _(null), d = _(null);
  return Q(async () => {
    const n = I(t);
    if (n) {
      if (D(n)) {
        l.value = n;
        return;
      }
    } else return;
    try {
      d.value = X(
        n
      ), c && (d.value.onPassword = (s, u) => {
        c({
          callback: s,
          isWrongPassword: u === Y.INCORRECT_PASSWORD
        });
      }), r && (d.value.onProgress = r), l.value = await d.value.promise;
    } catch (s) {
      if (l.value = null, o)
        o(s);
      else
        throw s;
    }
  }), B(l, (n, s) => {
    s == null || s.destroy();
  }), H(() => {
    var n, s, u, p;
    (n = d.value) != null && n.onPassword && (d.value.onPassword = null), (s = d.value) != null && s.onProgress && (d.value.onProgress = null), (u = d.value) == null || u.destroy(), D(I(t)) || (p = l.value) == null || p.destroy();
  }), {
    doc: l,
    download: async (n) => {
      if (!l.value)
        return;
      const s = await l.value.getData(), u = await l.value.getMetadata(), p = (
        // @ts-expect-error: contentDispositionFilename is not typed
        n ?? u.contentDispositionFilename ?? ""
      );
      de(s, p);
    },
    print: async (n = 300, s = "", u) => {
      var R, S, e, i;
      if (!l.value)
        return;
      const p = n / 72, x = 96 / 72;
      let f, w, P;
      try {
        f = window.document.createElement("div"), f.style.display = "none", window.document.body.appendChild(f), w = await le(f);
        const a = Math.max(3, Math.floor(10 * (300 / n))), y = u ? Array.isArray(u) ? u : [u] : [...Array(l.value.numPages + 1).keys()].slice(1);
        for (let m = 0; m < y.length; m += a)
          await Promise.all(
            y.slice(m, m + a).map(async (W, T) => {
              const k = await l.value.getPage(W), v = k.getViewport({
                scale: 1,
                rotation: 0
              });
              if (m + T === 0) {
                const M = v.width * p / x, U = v.height * p / x;
                re(w, M, U);
              }
              const h = window.document.createElement("canvas");
              h.width = v.width * p, h.height = v.height * p, f.appendChild(h);
              const L = h.cloneNode();
              w.contentWindow.document.body.appendChild(L), await k.render({
                canvasContext: h.getContext("2d"),
                intent: "print",
                transform: [p, 0, 0, p, 0, 0],
                viewport: v
              }).promise, L.getContext("2d").drawImage(h, 0, 0), N(h);
            })
          );
        s && (P = window.document.title, window.document.title = s), (R = w.contentWindow) == null || R.focus(), (S = w.contentWindow) == null || S.print();
      } finally {
        P && (window.document.title = P), V((e = w.contentWindow) == null ? void 0 : e.document.body), V(f), (i = f.parentNode) == null || i.removeChild(f);
      }
    }
  };
}
const ue = ["id"], pe = ["id"], he = {
  key: 0,
  class: "textLayer"
}, fe = {
  key: 1,
  class: "annotationLayer"
}, ge = /* @__PURE__ */ Z({
  __name: "VuePdfEmbed",
  props: {
    annotationLayer: { type: Boolean },
    height: {},
    id: {},
    imageResourcesPath: {},
    linkService: {},
    page: {},
    rotation: { default: 0 },
    scale: { default: 1 },
    source: {},
    textLayer: { type: Boolean },
    width: {}
  },
  emits: ["internal-link-clicked", "loaded", "loading-failed", "password-requested", "progress", "rendered", "rendering-failed"],
  setup(o, { expose: c, emit: r }) {
    const t = o, l = r, d = _([]), b = ee([]), g = _(null);
    let n = null;
    const { doc: s, download: u, print: p } = ce({
      onError: (e) => {
        d.value = [], l("loading-failed", e);
      },
      onPasswordRequest({ callback: e, isWrongPassword: i }) {
        l("password-requested", { callback: e, isWrongPassword: i });
      },
      onProgress: (e) => {
        l("progress", e);
      },
      source: te(t, "source")
    }), x = ae(() => {
      if (!s.value || !t.annotationLayer)
        return null;
      if (t.linkService)
        return t.linkService;
      const e = new se();
      return e.setDocument(s.value), e.setViewer({
        scrollPageIntoView: ({ pageNumber: i }) => {
          l("internal-link-clicked", i);
        }
      }), e;
    }), f = (e) => {
      let i, a;
      return t.height && !t.width ? (a = t.height, i = a / e) : (i = t.width ?? g.value.clientWidth, a = i * e), [i, a];
    }, w = async () => {
      if (!(!s.value || n != null && n.isAborted))
        try {
          d.value = t.page ? Array.isArray(t.page) ? t.page : [t.page] : [...Array(s.value.numPages + 1).keys()].slice(1), b.value = Array(d.value.length).fill(1), await Promise.all(
            d.value.map(async (e, i) => {
              const a = await s.value.getPage(e);
              if (n != null && n.isAborted)
                return;
              const y = ((t.rotation % 90 === 0 ? t.rotation : 0) + a.rotate) % 360, [m, W, T] = Array.from(
                g.value.getElementsByClassName("vue-pdf-embed__page")[i].children
              ), k = !!(y / 90 % 2), v = a.view[2] - a.view[0], h = a.view[3] - a.view[1], [L, M] = f(
                k ? v / h : h / v
              ), U = `${Math.floor(L)}px`, G = `${Math.floor(M)}px`, F = L / (k ? h : v), E = a.getViewport({
                scale: F,
                rotation: y
              });
              b.value[i] = F, m.style.display = "block", m.style.width = U, m.style.height = G;
              const $ = [
                P(
                  a,
                  E.clone({
                    scale: E.scale * window.devicePixelRatio * t.scale
                  }),
                  m
                )
              ];
              return t.textLayer && $.push(
                S(
                  a,
                  E.clone({
                    dontFlip: !0
                  }),
                  W
                )
              ), t.annotationLayer && $.push(
                R(
                  a,
                  E.clone({
                    dontFlip: !0
                  }),
                  T || W
                )
              ), Promise.all($);
            })
          ), n != null && n.isAborted || l("rendered");
        } catch (e) {
          d.value = [], b.value = [], n != null && n.isAborted || l("rendering-failed", e);
        }
    }, P = async (e, i, a) => {
      a.width = i.width, a.height = i.height, await e.render({
        canvasContext: a.getContext("2d"),
        viewport: i
      }).promise;
    }, R = async (e, i, a) => {
      q(a), await new J({
        accessibilityManager: null,
        annotationCanvasMap: null,
        annotationEditorUIManager: null,
        div: a,
        page: e,
        structTreeLayer: null,
        viewport: i
      }).render({
        annotations: await e.getAnnotations(),
        div: a,
        imageResourcesPath: t.imageResourcesPath,
        linkService: x.value,
        page: e,
        renderForms: !1,
        viewport: i
      });
    }, S = async (e, i, a) => {
      q(a), await new K({
        container: a,
        textContentSource: await e.getTextContent(),
        viewport: i
      }).render();
    };
    return B(
      s,
      (e) => {
        e && l("loaded", e);
      },
      { immediate: !0 }
    ), B(
      () => [
        s.value,
        t.annotationLayer,
        t.height,
        t.imageResourcesPath,
        t.page,
        t.rotation,
        t.scale,
        t.textLayer,
        t.width
      ],
      async ([e]) => {
        e && (n && (n.isAborted = !0, await n.promise), V(g.value), n = {
          isAborted: !1,
          promise: w()
        }, await n.promise, n = null);
      },
      { immediate: !0 }
    ), H(() => {
      V(g.value);
    }), c({
      doc: s,
      download: u,
      print: (e, i, a = !1) => p(e, i, a ? void 0 : t.page)
    }), (e, i) => (C(), A("div", {
      id: e.id,
      ref_key: "root",
      ref: g,
      class: "vue-pdf-embed"
    }, [
      (C(!0), A(ne, null, oe(d.value, (a, y) => (C(), A("div", { key: a }, [
        O(e.$slots, "before-page", { page: a }),
        z("div", {
          id: e.id && `${e.id}-${a}`,
          class: "vue-pdf-embed__page",
          style: ie({
            "--scale-factor": b.value[y],
            position: "relative"
          })
        }, [
          i[0] || (i[0] = z("canvas", null, null, -1)),
          e.textLayer ? (C(), A("div", he)) : j("", !0),
          e.annotationLayer ? (C(), A("div", fe)) : j("", !0)
        ], 12, pe),
        O(e.$slots, "after-page", { page: a })
      ]))), 128))
    ], 8, ue));
  }
});
export {
  ke as GlobalWorkerOptions,
  ge as default,
  ce as useVuePdfEmbed
};
